common4j/src/main/com/microsoft/identity/common/java/crypto/StorageEncryptionManager.java

```diff
@@ -0,0 +1,511 @@
+                                              final int encodeVersionLength) throws ClientException {
+        if (encodeVersionLength <= 0) {
+            throw new ClientException(
+                    DATA_MALFORMED,
+                    String.format(
+                            "Encode version length: '%s' is not valid, it must be greater of equal to 0",
+                            encodeVersionLength
+                    )
+            );
+        }
+
+        if (encodeVersionLength + 1 > cipherString.length()){
+            throw new ClientException(
+                    DATA_MALFORMED,
+                    "Length of encode version string (plus the length character) is longer than " +
+                            "the CipherString itself. The data is malformed.");
+        }
+
+        if (!cipherString.substring(1, 1 + encodeVersionLength).equals(ENCODE_VERSION)) {
+            throw new ClientException(
+                    DATA_MALFORMED,
+                    String.format(
+                            "Unsupported encode version received. Encode version supported is: '%s'",
+                            ENCODE_VERSION
+                    )
+            );
+        }
+    }
+
+    /**
+     * Compare digest of input message and calculated MAC digest.
+     *
+     * @param start    Start index of the digest in the encrypted blob array.
+     * @param end      End index of the digest in the encrypted blob array.
+     * @param expected The expected mac digest.
+     * @throws ClientException Client if there is any mismatch.
+     */
+    private void assertHMac(final byte[] encryptedBlob,
+                            final int start,
+                            final int end,
+                            final byte[] expected)
+            throws ClientException {
+        if (expected.length != (end - start)) { //NOPMD
+            throw new ClientException(UNEXPECTED_HMAC_LENGTH);
+        }
+
+        byte result = 0;
+        // It does not fail fast on the first not equal byte to protect against
+        // timing attack.
+        for (int i = start; i < end; i++) {
+            result |= expected[i - start] ^ encryptedBlob[i];
+        }
+
+        if (result != 0) {
+            throw new ClientException(HMAC_MISMATCH);
+        }
+    }
+
+    /**
+     * Get Secret Key to use in encryption/decryption.
+     *
+     * @return a SecretKey loader.
+     */
+    @NonNull
+    public abstract AbstractSecretKeyLoader getKeyLoaderForEncryption() throws ClientException;
+
+    /**
+     * Identify the encrypted blob and return a list of potential candidate key loaders for decryption.
+     *
+     * @param cipherText The cipherText to decrypt.
+     * @return a prioritized list of SecretKey (earlier keys is more likely to be the correct one).
+     **/
+    @NonNull
+    abstract public List<AbstractSecretKeyLoader> getKeyLoaderForDecryption(@NonNull final byte[] cipherText) throws ClientException;
```

Comment:
Do we necessarily need this? Instead of having these `abstract` method as a means of defining "Key Loaders", would it better to just directly accept the `SecretKey` as an argument to the constructor of this class? The way I see it is that this class doesn't necessarily need to "load a key"..it just needs to know the key that needs to be used for encryption / decryption.

Reply:
1. KeyLoaders also provides metadata (i.e. crypto algorithm) related to the key, such as 
   - we're using AES/CBC/PKCS5Padding for AES-256 key
   - We use "A001" as an identifier for Keystore-wrapped key, and "U001" for predefined key.

Reply:
I believe we can the algorithm by doing `mSecretKey.getAlgorithm()`, right? Or is that different? 

For key identifier we will probably have to pass it in the constructor (if we choose to go that route).

Reply:
mSecretKey.getAlgorithm() -> you'll get "AES" (algo for the key itself)

AES/CBC/PKCS5Padding is what we pass into Cipher.getInstance(). I might have used the wrong term here. Maybe a better name for it is *transformation*.

Reply:
by constructor, which one are you specifically referring to? We can't pass it in this class because the identifier is tied to the 'category' of key being used.

Reply:
I think between `KeyLoader` vs the `SecretKey`, I think both patterns are okay. I guess what I'm trying to understand is why these need to be `abstract` methods. What about supplying the `KeyLoader` directly into the constructor for `StorageEncryptionManager`? I think both are fine...just curios if there is a reason for picking abstract methods here.

Reply:
Or Maybe "Key Loader" is a wrong term here?

Reply:
Broker loads the key one way. 
ADAL/MSAL loads the key another way.

(different set/order of keys)

There are two ways to do this, either abstract it out, or check+inject the keys everywhere, and I went with the first one.

Reply:
(That said, once we finish the work e2e, then the key picking logic can reside in XXXXComponent, but i wouldn't want to bloat that class either).

Reply:
Well...  i could make factory classes out of those.

Reply:
Decide to keep it like this for now.